### PR TITLE
SQP-basd Trajectory optimization

### PR DESCRIPTION
**Pull request by HiroIshida**
### feature
SQP-based trajectory optimization. The cost of a trajecotory is a CHOMP-type one (quadratic cost), but this implementation is different from CHOMP. CHOMP use an objective-function that is the sum of trajectory cost and collision cost and then minimize it by grad-based method. On the other hand, in this implementation, collision cost is directly considered as an inequality constrain.

Please try `example/collision_free_trajectory.py`. Assuming my previous PR https://github.com/iory/scikit-robot/pull/173 works correctly, this GIF shows that there is no collision between the sphere and the object throughout the trajectory. 

out

### warning 
Note that grad-based trajecotyr optimizer can be easily got stuck into a local optima. If you make the box tinier, then the trajectory will tend to jump over the object. So, please consider that this optimization based method can be mainly applied to an easy problem. 

For difficult problem, sampling based solver based initialization is necessary. But, current FK solving speed of scikit-robot, it will take several minutes, So it's not practical. 

### TODO
finnished!
- [x] implementation
- [x] example
- [x] test 
- [x] documentation
- [x] clean up

### CODE REVIEW

File: skrobot/planner/sqp_based.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
+import numpy as np
+from repoze.lru import lru_cache
```
Comment (iory) [inline skrobot/planner/sqp_based.py:2]: In python3, is `functools.lru_cache` faster than `repoze.lru.lru_cache`?
https://gist.github.com/iory/e82dab918c0a5c32cd3b1a4d1c06adbf

Comment (HiroIshida) [reply skrobot/planner/sqp_based.py]: Thank you for your time to giving me a benchmark. I fixed it in this commit :https://github.com/iory/scikit-robot/pull/181/commits/27bde96bea67cbe46312f3f4297e353c43fe035e

File: requirements.txt [hunk 2]
```diff
@@ -19,3 +19,4 @@ six
 trimesh>=3.5.20
 sphinx==1.8.2
 sphinx_rtd_theme
+repoze.lru
```
Comment (iory) [inline requirements.txt:22]: It needs only for python2.7. So rewrite it to `repoze.lru;python_version<"3.2"`.

**Review by iory — CHANGES_REQUESTED** [review]

File: skrobot/planner/sqp_based.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    joint_list : list[skrobot.model.Link]
+        link list to be controlled (similar to inverse_kinematics function)
+    n_wp : int
+        number of waypoints
+    safety_margin : float
+        safety margin in collision checking
+    with_base: bool
+        If `with_base=False`, `n_dof` is the number of joints `n_joint`,
+        but if `with_base=True`, `n_dof = len(joint_list) + 3`.
+    weights : Union[numpy.array(n_dof,), None]
```
Comment (iory) [inline skrobot/planner/sqp_based.py:46]: What is the Union?

File: skrobot/planner/sqp_based.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+        cost to move of each joint. For example,
+        if you set weights=numpy.array([1.0, 0.1, 0.1]) for a
+        3 DOF manipulator, moving the first joint is with
+        high cost compared to others. If set to `None` it's automatically
+        determined.
+    initial_trajectory : Union[numpy.array(n_wp, n_dof), None]
+        initial solution in the trajectory optimization specifeid by a
+        angle vector sequence. If None, initial trajectory is
+        automatically generated.
+    slsqp_option: Union[Dict, None]
```
Comment (iory) [inline skrobot/planner/sqp_based.py:56]: This `Dict` shoule be `dict.`

**Review by iory — CHANGES_REQUESTED** [review]

File: skrobot/planner/sqp_based.py (hunk 3 above)
Comment (HiroIshida) [reply skrobot/planner/sqp_based.py:46]: following this : https://stackoverflow.com/questions/34647966/how-to-express-multiple-types-for-a-single-parameter-or-a-return-value-in-docstr

**Comment by iory** [comment]
Some of docstring's numpy's link is broken. 
If you can write it correctly, numpy.ndarray's links are automatically generated as follow.
https://scikit-robot.readthedocs.io/en/latest/reference/generated/skrobot.coordinates.math.inverse_rodrigues.html#skrobot.coordinates.math.inverse_rodrigues

File: skrobot/planner/sqp_based.py (hunk 4 above)
Comment (HiroIshida) [reply skrobot/planner/sqp_based.py]: fixed in https://github.com/iory/scikit-robot/pull/181/commits/25cc3d02aa0309727d461758d0bb62043a73a95e

File: requirements.txt (hunk 2 above)
Comment (HiroIshida) [reply requirements.txt]: fixed in https://github.com/iory/scikit-robot/pull/181/commits/25cc3d02aa0309727d461758d0bb62043a73a95e

File: skrobot/planner/sqp_based.py (hunk 3 above)
Comment (iory) [reply skrobot/planner/sqp_based.py:46]: We are using numpy style for docstrings. `numpy.ndarray or None` is suitable.

Comment (HiroIshida) [reply skrobot/planner/sqp_based.py]: fixed in https://github.com/iory/scikit-robot/pull/181/commits/66a1bd946c7bfeb1cda94b7bf780461877b62d16

**Review by iory — APPROVED** [review]

**Comment by HiroIshida** [comment]
cleaned up the commit logs and now ready to be merged!

**Comment by iory** [comment]
LGTM!

**Pull request merged**